consensus/core/src/block.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    cell::OnceCell,
+    hash::{Hash, Hasher},
+};
+
+use fastcrypto::hash::HashFunction;
+use serde::{Deserialize, Serialize};
+
+use consensus_config::{AuthorityIndex, DefaultHashFunction};
+
+/// Round number of a block.
+pub type Round = u32;
+
+/// Block proposal timestamp in milliseconds.
+pub type BlockTimestampMs = u64;
+
+/// One validator can produce at most one block per round.
+/// A block includes references to previous round blocks and transactions that the validator
+/// considers valid.
+#[derive(Clone, Deserialize, Serialize)]
+pub enum Block {
+    V1(BlockV1),
+}
+
+pub trait BlockAPI {
+    fn reference(&self) -> BlockRef;
+    fn digest(&self) -> BlockDigest;
+    fn round(&self) -> Round;
+    fn author(&self) -> AuthorityIndex;
+    fn timestamp(&self) -> BlockTimestampMs;
+    fn includes(&self) -> &[BlockRef];
+    // TODO: add accessor for transactions.
+}
+
+#[derive(Clone, Default, Deserialize, Serialize)]
+pub struct BlockV1 {
+    round: Round,
+    author: AuthorityIndex,
+    timestamp: BlockTimestampMs,
+    includes: Vec<BlockRef>,
+
+    #[serde(skip)]
+    digest: OnceCell<BlockDigest>,
+}
+
+impl BlockAPI for BlockV1 {
+    fn reference(&self) -> BlockRef {
+        BlockRef {
+            round: self.round,
+            author: self.author,
+            digest: self.digest(),
+        }
+    }
+
+    fn digest(&self) -> BlockDigest {
+        *self.digest.get_or_init(|| {
+            let mut hasher = DefaultHashFunction::new();
+            hasher.update(bcs::to_bytes(&self).expect("Serialization should not fail"));
+            BlockDigest(hasher.finalize().into())
+        })
```

Comment:
Maybe let's directly implement the hash calculation under the implementation of the Hash trait ? Ex
```
impl Hash<{ crypto::DIGEST_LENGTH }> for BlockV1 {
    type TypedDigest = BlockDigest;

    fn digest(&self) -> BlockDigest {
        let mut hasher = DefaultHashFunction::new();
        hasher.update(bcs::to_bytes(&self).expect("Serialization should not fail"));
        BlockDigest(hasher.finalize().into())
    }
}
```

Reply:
I'm see the `fastcrypto::hash::Hash` trait implemented for most Narwhal types, but I'm not sure where it is used .. Also I think caching the digest is preferrable. So implementing the trait only for `enum Block`.